language/move-lang/src/interface_generator.rs

```diff
@@ -0,0 +1,284 @@
+    counts: BTreeMap<String, usize>,
+}
+
+impl<'a> Context<'a> {
+    fn new(module: &'a CompiledModule) -> Self {
+        Self {
+            module,
+            uses: BTreeMap::new(),
+            counts: BTreeMap::new(),
+        }
+    }
+}
+
+const DISCLAIMER: &str = "// NOTE: Functions are 'native' for simplicity. They may or \
+                          may not be native in actuality.";
+
+fn write_struct_def(ctx: &mut Context, sdef: &StructDefinition) -> String {
+    let mut out = String::new();
+
+    let shandle = ctx.module.struct_handle_at(sdef.struct_handle);
+    let resource_mod = if shandle.is_nominal_resource {
+        "resource "
+    } else {
+        ""
+    };
+
+    push_line!(
+        out,
+        format!(
+            "    {}struct {}{} {{",
+            resource_mod,
+            ctx.module.identifier_at(shandle.name),
+            write_type_paramters(&shandle.type_parameters),
+        )
+    );
+
+    let fields = match &sdef.field_information {
+        StructFieldInformation::Native => {
+            push!(out, "    }");
+            return out;
+        }
+        StructFieldInformation::Declared(fields) => fields,
+    };
+    for field in fields {
+        push_line!(
+            out,
+            format!(
+                "        {}: {},",
+                ctx.module.identifier_at(field.name),
+                write_signature_token(ctx, &field.signature.0),
+            )
+        )
+    }
+
+    push!(out, "    }");
+    out
+}
+
+fn write_function_def(ctx: &mut Context, fdef: &FunctionDefinition) -> String {
+    let fhandle = ctx.module.function_handle_at(fdef.function);
+    let parameters = &ctx.module.signature_at(fhandle.parameters).0;
+    let return_ = &ctx.module.signature_at(fhandle.return_).0;
+    format!(
+        "    native public fun {}{}({}){};",
+        ctx.module.identifier_at(fhandle.name),
+        write_type_paramters(&fhandle.type_parameters),
+        write_parameters(ctx, parameters),
+        write_return_type(ctx, return_)
+    )
+}
+
+fn write_type_paramters(tps: &[Kind]) -> String {
+    if tps.len() == 0 {
+        return "".to_string();
+    }
+
+    let tp_and_constraints = tps
+        .iter()
+        .enumerate()
+        .map(|(idx, kind)| {
+            format!(
+                "{}{}",
+                write_type_parameter(idx as TypeParameterIndex),
+                write_kind_contraint(kind)
+            )
+        })
+        .collect::<Vec<_>>()
+        .join(", ");
+    format!("<{}>", tp_and_constraints)
+}
+
+fn write_kind_contraint(kind: &Kind) -> String {
+    match kind {
+        Kind::All => "".to_string(),
+        Kind::Resource => ": resource".to_string(),
+        Kind::Copyable => ": copyable".to_string(),
+    }
+}
+
+fn write_parameters(ctx: &mut Context, params: &[SignatureToken]) -> String {
+    params
+        .iter()
+        .enumerate()
+        .map(|(idx, ty)| format!("a{}: {}", idx, write_signature_token(ctx, ty)))
```

Comment:
Our current move code uses capital letters (e.g. `T`, `CoinType`). Do we also want to follow that convention here?